crates/sui-core/src/authority/authority_store.rs

```diff
@@ -540,6 +541,93 @@ impl AuthorityStore {
+            epoch_id,
+            ObjectKey(*object_id, version),
+            MarkerKind::ReceivedDeletedOrWrappedAtVersion,
+        );
+        Ok(self
+            .perpetual_tables
+            .object_per_epoch_marker_table
+            .get(&marker_key)?
+            .is_some())
+    }
+
+    pub fn have_received_deleted_or_wrapped_object_at_version_or_after(
+        &self,
+        object_id: &ObjectID,
+        version: VersionNumber,
+        epoch_id: EpochId,
+    ) -> Result<bool, SuiError> {
+        let object_key = ObjectKey::max_for_id(object_id);
+        let marker_key = (
+            epoch_id,
+            object_key,
+            MarkerKind::ReceivedDeletedOrWrappedAtVersion,
+        );
+
+        // Find the most recent version of the object that was received, deleted or wrapped.
+        // Return true if the version is >= `version`. Otherwise return false.
+        let marker_entry = self
+            .perpetual_tables
+            .object_per_epoch_marker_table
+            .unbounded_iter()
+            .skip_prior_to(&marker_key)?
+            .next();
+        match marker_entry {
+            Some(((_, key, _), _)) => Ok(key.0 == *object_id && key.1 >= version),
+            None => Ok(false),
+        }
+    }
+
+    pub fn check_receiving_objects(
+        &self,
+        receiving_objects: &[ObjectRef],
+        input_objects_len: usize,
+        epoch_store: &AuthorityPerEpochStore,
+    ) -> Result<(), SuiError> {
+        let protocol_config = epoch_store.protocol_config();
+        // Count receiving objects towards the input object limit as they are passed in the PTB
+        // args and they will (most likely) incur an object load at runtime.
+        fp_ensure!(
+            receiving_objects.len() + input_objects_len
+                <= protocol_config.max_input_objects() as usize,
+            UserInputError::SizeLimitExceeded {
+                limit: "maximum input and receiving objects in a transaction".to_string(),
+                value: protocol_config.max_input_objects().to_string()
+            }
+            .into()
+        );
+
+        // Since we're at signing we check that every object reference that we are receiving is the
+        // most recent version of that object. If it's been received at the version specified we
+        // let it through to allow the transaction to run and fail to unlock any other objects in
+        // the transaction. Otherwise, we return an error.
+        for (object_id, version, _) in receiving_objects {
+            fp_ensure!(
+                self.get_object(object_id)?
+                    .is_some_and(|x| x.owner.is_address_owned() && x.version() == *version)
+                    || self.have_received_deleted_or_wrapped_object_at_version(
+                        object_id,
+                        *version,
+                        epoch_store.epoch()
+                    )?,
+                UserInputError::ObjectNotFound {
```

Comment:
Good point, I'll make a pass through here to give better/more specific errors depending on the exact type of error.

Reply:
Updated this to give the correct error depending on the error (e.g., invalid seqno, mismatching digest, shared object instead of owned, package, immutable), and added a test to verify we give the correct error in each case.